src/Mvc/src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -103,6 +115,40 @@ public TestServer Server
         public WebApplicationFactory<TEntryPoint> WithWebHostBuilder(Action<IWebHostBuilder> configuration) =>
             WithWebHostBuilderCore(configuration);
 
+        /// <summary>
+        /// Gets the content root path for the <see cref="WebApplicationFactory{TEntryPoint}"/>, if known.
+        /// </summary>
+        /// <returns>
+        /// The path of the content root path to use, if known; otherwise <see langword="null"/>.
+        /// </returns>
+        protected string GetContentRootPath()
```

Comment:
Can you provide a specific use-case for this? I don't want to add public APIs if we don't have to as these are more things to support/maintain.

Could you not get the ContentRoot from the IHost/IWebHost directly?

Reply:
The default is being calculated incorrectly in the use case where I'm deriving from `WAF<T>` to host an HTTP server on a real port for Selenium UI tests, so there's no point reaching in for it, because it's wrong. This mean that the application was failing to find various static assets, like JS files, so the tests were failing.

I needed to manually "fix" it, and that involved having to duplicate this code so I could re-use it:

https://github.com/martincostello/alexa-london-travel-site/blob/e9af0ce3e70c7cbb0d047a276db42491ccec89e5/tests/LondonTravel.Site.Tests/Integration/TestServerFixture.cs#L83

It's possible that there's a separate issue either here or in my code that means the root is being set/calculated/overridden incorrectly in the first place, but it seemed "better" to be able to just leverage the code that was already there (but privately), rather than having to copy-paste as I've had to do with 2.2 and 3.0 preview 2.

Reply:
@martincostello The WAF is meant to be used specifically for in-memory integration tests. 
I want to make sure you understand that there is currently no support for standing up a real server using the web application factory. 

I want to be very careful in exposing the guts of the WAF if there is no good reason for it, as each time we add public API it has a cost for us and limits our options in the future. The WAF was designed to enable in-memory integration tests and trying to use it for E2E tests I believe doesn't have a lot of benefits compared to copying the same approach we follow in other places (in our actual E2E tests).

If you are trying to do E2E tests with Selenium we have other things that might help in this area, although nothing is official, you can follow the same approach. See
https://github.com/aspnet/AspNetCore/tree/master/src/Hosting/Server.IntegrationTesting
https://github.com/aspnet/AspNetCore/tree/master/src/Middleware/CORS/test/FunctionalTests
https://github.com/aspnet/AspNetCore/blob/master/src/ProjectTemplates/test/Infrastructure/SeleniumServerFixture.cs

I would be interested in getting a better feeling on what specifically made you want to use the WebApplicationFactory along with Selenium and see if there is a way we can factor those things out for ease of use instead of exposing the WAF guts.

Reply:
The main motivation was because WAF contains a lot of boilerplate to make sure that Razor compilation references work correctly for self-hosted tests on a non-deployed application.

Previously trying to do this without leveraging WAF just meant that the application failed to find the references when Razor views were executed, which broke the application.

As Selenium tests to do UI automation (where buttons need to be clicked, etc.), an HTTP port has to be exposed so that the browser can actually interact with the application. This sort of testing isn't possible in memory.

In memory works great for APIs and very basic HTML applications, but it doesn't work for more complex HTML-based applications where you want to do as much testing in the inner loop as possible without also having to maintain a lot of extra boilerplate to make it work.

WAF happens to contain most of this boilerplate, but due to the design being through inheritance and protected methods, rather than generic callable public APIs, the only way to leverage it at the moment to to create derived types and then customise the behaviour in the derived type.

This PR has come about from trying to port existing web applications using this approach to use 3.0 in advance of its release where they didn't work properly anymore because of the hosting changes.

Reply:
If it's useful to you, having done some refactoring in my master branch already to make relevant updates for 2.2.x to keep the diff size down, a branch with the changes I've had to make to one of my applications that illustrates the workarounds I've used with preview 2 to keep things working with 3.0 is here: https://github.com/martincostello/alexa-london-travel-site/pull/255/files